File: leetcode/62.py
# 62. Unique Paths
# Solution 1
# Time Complexity : O(m*n)
# Using dp, dp[i][j] = dp[i-1][j]+dp[i][j]
# b/c it is always unique path 

class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
     	dp = []
	# Making dp table
	for i in range(m):
		tmp = []
		# Initialization
		for j in range(n): 
			if i == 0 :
				tmp.append(1)
			elif j == 0 :
				tmp.append(1)
			else : 
				tmp.append(0)
		dp.append(tmp)
        
	for i in range(1,n):
            for j in range(1,m):
		dp[i][j] = dp[i-1][j]+dp[i][j-1]
	return dp[n-1][m-1]

File: leetcode/62.cpp
// 62. Unique Paths 


// Making DP table  int [][] - Beat 100%
class Solution {
public:
    int uniquePaths(int m, int n) {
        /*int ** dp;
        // initialize dp table 
        dp = new int *[n];
        for(int i=0; i<n; i++)
            dp[i] = new int[m];*/
        
        
        int dp[100][100];
        
        for(int i = 0; i<m; i++)
            dp[0][i] = 1;
        for(int i=0; i<n; i++)
            dp[i][0] = 1;
        for(int i=1; i<n; i++){
            for(int j=1; j<m; j++){
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
        
        return dp[n-1][m-1];
    }
};
